pca-distances.1
.\" -*- nroff -*-
.ds g \" empty
.ds G \" empty
.de Tp
.ie \\n(.$=0:((0\\$1)*2u>(\\n(.1u-\\n(.iu)) .TP
.el .TP "\\$1"
..
.TH PCA-DISTANCES 1 "13 Jul 2015" "pca-utils version 20150713"
.SH NAME
pca-distances \- build distance matrices from scores using various metrics
.SH SYNOPSIS
.B pca-distances
[
.B \-h
]
.LP
.B pca-distances
\-i \fIINFILE\fR [\fIOPTIONS\fR]
.SH DESCRIPTION
.PP
Build a distance matrix from PCA/PLS scores in \fBINFILE\fR. The matrix can
be built using various distance metrics defined below.
.TP
\fB\-h\fR, \fB-\-help\fR
display a short help message.
.TP
\fB\-i\fR, \fB\-\-input\fR \fIINFILE\fR
scores list input file, formatted according to the list file exported from
Umetrics SIMCA-P+ software. See pca-rand(1) for more information.
.TP
\fB\-m\fR, \fB\-\-metric\fR \fIMETRIC\fR
distance metric to use. Supported metrics are defined below.
.TP
\fB\-D\fR, \fB\-\-divisions\fR \fINDIV\fR
number of grid divisions to create while approximating probability densities
for grid-based distance metrics.
.SH "DISTANCE METRICS"
.TP
.B EUC: Euclidean
the square root of the dot product of the displacement vector between the
two group means. special case of the mahalanobis distance that assumes a
covariance matrix equal to the identity.
.TP
.B MAH: Mahalanobis
closely related to the square root of the exponent of a multivariate normal
density function. the covariance matrix used to weight the displacement is
a weighted sum of the covariance matrices of each group; this implies the
assumption that the groups are nearly identically distributed, an
approximation that works well in practice.
.TP
.B BHA: Bhattacharyya
related to the mahalanobis distance, but makes no assumption of identically
distributed groups. this metric, however, does not obey the triangle
inequality. this assumes normally distributed points in each group.
.TP
.B HEL: Hellinger
a variant of the bhattacharyya distance that uses the same overlap
coefficient, but forms a metric with it differently, thus obeying the
triangle inequality.
.TP
.B MIX: Mixture
approximated bhattacharyya distance for two-dimensional scores only, which
lays a grid onto the scores space and convolves each point as a unit impulse
against a gaussian point spread function.
.TP
.B CHI: Chi-square
uses a similar means to approximate density functions as the mixture metric,
but attempts to calculate a chi-square statistic from the gridded data. this
fails in practice due to the large number of zero-valued bins in the matrix.
.SH AUTHOR
Bradley Worley <geekysuavo@gmail.com>
.SH COPYRIGHT
Copyright \(co 2015 Bradley Worley.
.br
This is free software. You may redistribute copies of it under the terms of
the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR pca-bootstrap(1),
.BR pca-dendrogram(1),
.BR pca-ellipses(1),
.BR pca-ellipsoids(1),
.BR pca-overlap(1),
.BR pca-stats(1),
.BR pca-rand(1),
.BR pca-maps(1)


pca-distances.c

/*
 * pca-distances.c: main program source code for pca-distances.
 * Copyright (C) 2015 Bradley Worley <geekysuavo@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/* include the pca utils header. */
#include "pca-utils.h"

/* define the help message string. */
#define HELP "\
 pca-distances: A tool to build distance matrices of PCA group scores.\n\
 Copyright (C) 2015 Bradley Worley. Released under the GNU GPL 3.0.\n\
\n\
 Usage:\n\
   pca-distances [arguments]\n\
\n\
 Arguments:\n\
   -h,--help            Display this help message\n\
   -i,--input FIN       Input scores list filename (required)\n\
   -m,--metric DM       One of the supported distance metrics [EUC]\n\
   -D,--divisions DIV   Number of divisions for MIX, CHI only [128]\n\
\n\
 Metrics:\n\
   EUC: Euclidean\n\
        This is simply the square root of the dot product of the vector\n\
        of displacements between the group means.\n\
   MAH: Mahalanobis\n\
        This is a weighted metric, where the inverse of the pooled\n\
        variance-covariance matrix of the two groups is placed into\n\
        the multiplication.\n\
   BHA: Bhattacharyya\n\
        This is another weighted metric, based on the negative natural\n\
        logarithm of the Bhattacharyya coefficient for the two groups.\n\
   HEL: Hellinger\n\
        Similar to the Bhattacharyya distance, this is the square root\n\
        of one minus the Bhattacharyya coefficient. Unlike BHA, HEL obeys\n\
        the triangle inequality.\n\
   MIX: Mixture\n\
        Builds a Bhattacharyya distance by approximating the overlap\n\
        integral between groups in 2D space using Riemann sums. The\n\
        probability density of each group is estimated by convolving\n\
        every point in the group with a Gaussian point spread function\n\
        and normalizing to unit volume.\n\
   CHI: Chi-square\n\
        Using the same convolution as MIX, this metric calculates a chi-\n\
        square value for each pair of groups. Note that this value will\n\
        not necessarily vary as a chi-square distribution, since the\n\
        vast majority of compared pixels will be zero.\n\
\n\
 The pca-distances tool reads in an input PCA scores file and generates a\n\
 symmetric matrix of intergroup distances, using the distance metric DM,\n\
 of which it prints only the upper triangle, with nice group labels.\n\
\n\
"

/* main: application entry point.
 * @argc: the number of commandline arguments.
 * @argv: the commandline argument string array.
 */
int main (int argc, char **argv) {
  /* declare required variables. */
  pca_distance_metric metric;
  char *ifname, *metstr;
  struct matrix *D;
  struct pca *P;

  /* parse the command-line arguments. */
  if (!opts_init (argc, argv)) {
    /* output an error message and exit. */
    ferror ("failed to parse arguments");
    fprintf (stdout, HELP);
    return 1;
  }

  /* see if we should display the help message. */
  if (opts_geti (OPTS_S_HELP)) {
    /* print the help message and quit. */
    fprintf (stdout, HELP);
    return 0;
  }

  /* get the input filename. */
  ifname = opts_gets (OPTS_S_INPUT);

  /* ensure the input filename was provided. */
  if (!ifname) {
    /* output an error message and exit. */
    ferror ("input file required");
    fprintf (stdout, HELP);
    return 1;
  }

  /* read in the input file. */
  P = pca_new (ifname);

  /* make sure we read in the file. */
  if (!P) {
    /* output an error message and exit. */
    ferror ("failed to read '%s'", ifname);
    return 1;
  }

  /* set the default distance metric. */
  metric = pca_group_distance_euclidean;

  /* get the distance metric string. */
  metstr = opts_gets (OPTS_S_METRIC);

  /* see if the distance metric was provided. */
  if (metstr) {
    /* check the distance metric argument. */
    if (strncmp (metstr, "EUC", 3) == 0 ||
        strncmp (metstr, "euc", 3) == 0) {
      /* set the euclidean metric. */
      metric = pca_group_distance_euclidean;
    }
    else if (strncmp (metstr, "MAH", 3) == 0 ||
             strncmp (metstr, "mah", 3) == 0) {
      /* set the mahalanobis distance. */
      metric = pca_group_distance_mahalanobis;
    }
    else if (strncmp (metstr, "BHA", 3) == 0 ||
             strncmp (metstr, "bha", 3) == 0) {
      /* set the bhattacharyya distance. */
      metric = pca_group_distance_bhattacharyya;
    }
    else if (strncmp (metstr, "HEL", 3) == 0 ||
             strncmp (metstr, "hel", 3) == 0) {
      /* set the hellinger distance. */
      metric = pca_group_distance_hellinger;
    }
    else if (strncmp (metstr, "MIX", 3) == 0 ||
             strncmp (metstr, "mix", 3) == 0) {
      /* set the mixture distance. */
      metric = pca_group_distance_mixture;
    }
    else if (strncmp (metstr, "CHI", 3) == 0 ||
             strncmp (metstr, "chi", 3) == 0) {
      /* set the mixture distance. */
      metric = pca_group_distance_chisq;
    }
    else {
      /* output an error message and exit. */
      ferror ("unrecognized distance metric '%s'", metstr);
      fprintf (stdout, HELP);
      return 1;
    }
  }

  /* allocate a distance matrix. */
  D = matrix_new (P->n_groups, P->n_groups);

  /* make sure we allocated a matrix. */
  if (!D) {
    /* output an error message and exit. */
    ferror ("failed to allocate distance matrix");
    return 1;
  }

  /* calculate the distance matrix. */
  if (!pca_distances (P, D, metric)) {
    /* output an error message and exit. */
    ferror ("failed to calculate distance matrix");
    return 1;
  }

  /* write out the distance matrix. */
  pca_distances_print (P, D);

  /* free the distance matrix. */
  matrix_free (D);

  /* free the pca structure. */
  pca_free (P);

  /* return success. */
  return 0;
}

